File: typing_trainer.go
// typing_trainer.go
package main

import (
	"bufio"
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"math/rand"
	"os"
	"strings"
	"time"
)

type Trainer struct {
	length   int
	charset  string
	mode     string
	language string
	text     string
	start    time.Time
	end      time.Time
	errors   int
	total    int
}

func NewTrainer(length int, charset, mode, language string) *Trainer {
	t := &Trainer{length: length, charset: charset, mode: mode, language: language}
	t.text = t.generateText()
	return t
}

func (t *Trainer) getCharset() string {
	var letters string
	if t.language == "ru" {
		letters = "абвгдеёжзийклмнопрстуфхцчшщъыьэюяАБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ"
	} else {
		letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
	}
	digits := "0123456789"
	symbols := "!@#$%^&*()_+-=,.;:?/[]{}|"
	switch t.charset {
	case "letters":
		return letters
	case "alnum":
		return letters + digits
	default:
		return letters + digits + symbols
	}
}

func (t *Trainer) generateText() string {
	charset := t.getCharset()
	if t.mode == "word" {
		count := len(charset)
		words := []string{}
		wordCount := max(1, t.length/5)
		for i := 0; i < wordCount; i++ {
			wl := rand.Intn(6) + 3
			word := ""
			for j := 0; j < wl; j++ {
				word += string(charset[rand.Intn(count)])
			}
			words = append(words, word)
		}
		return strings.Join(words, " ")
	} else if t.mode == "sentence" {
		sentences := []string{}
		sentenceCount := max(1, t.length/30)
		for i := 0; i < sentenceCount; i++ {
			wc := rand.Intn(8) + 5
			words := []string{}
			for j := 0; j < wc; j++ {
				wl := rand.Intn(6) + 3
				word := ""
				for k := 0; k < wl; k++ {
					word += string(charset[rand.Intn(len(charset))])
				}
				words = append(words, word)
			}
			sentence := strings.Join(words, " ")
			sentence = strings.ToUpper(sentence[:1]) + sentence[1:] + "."
			sentences = append(sentences, sentence)
		}
		return strings.Join(sentences, " ")
	} else {
		res := ""
		for i := 0; i < t.length; i++ {
			res += string(charset[rand.Intn(len(charset))])
		}
		return res
	}
}

func (t *Trainer) Run() map[string]interface{} {
	reader := bufio.NewReader(os.Stdin)
	fmt.Println("\033[36mТренажер печати (символы)\033[0m")
	fmt.Println("\033[33mВведите следующий текст как можно точнее и быстрее:\033[0m")
	fmt.Println("\033[37m" + t.text + "\033[0m")
	fmt.Println("\nНажмите Enter, чтобы начать...")
	reader.ReadString('\n')
	t.start = time.Now()
	fmt.Print("\033[32m> \033[0m")
	input, _ := reader.ReadString('\n')
	input = strings.TrimSuffix(input, "\n")
	t.end = time.Now()
	t.total = len(input)
	minLen := len(t.text)
	if len(input) < minLen {
		minLen = len(input)
	}
	t.errors = 0
	for i := 0; i < minLen; i++ {
		if t.text[i] != input[i] {
			t.errors++
		}
	}
	if len(input) > len(t.text) {
		t.errors += len(input) - len(t.text)
	}
	elapsed := t.end.Sub(t.start).Seconds()
	cpm := float64(t.total) / elapsed * 60
	wpm := cpm / 5
	accuracy := 0.0
	if t.total > 0 {
		accuracy = float64(t.total-t.errors) / float64(t.total) * 100
	}
	fmt.Println("\n\033[36m--- Результаты ---\033[0m")
	fmt.Printf("\033[32mСкорость: %.1f симв/мин (%.1f слов/мин)\033[0m\n", cpm, wpm)
	fmt.Printf("\033[33mТочность: %.1f%%\033[0m\n", accuracy)
	fmt.Printf("\033[35mВремя: %.2f сек\033[0m\n", elapsed)
	fmt.Printf("\033[31mОшибок: %d\033[0m\n", t.errors)
	return map[string]interface{}{
		"cpm": cpm, "wpm": wpm, "accuracy": accuracy, "time": elapsed, "errors": t.errors,
	}
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	var (
		length   int
		charset  string
		mode     string
		language string
		export   string
	)
	flag.IntVar(&length, "length", 50, "Длина текста")
	flag.StringVar(&charset, "charset", "all", "letters, alnum, all")
	flag.StringVar(&mode, "mode", "char", "char, word, sentence")
	flag.StringVar(&language, "language", "en", "en, ru")
	flag.StringVar(&export, "export-stats", "", "Экспорт статистики")
	flag.Parse()

	rand.Seed(time.Now().UnixNano())
	trainer := NewTrainer(length, charset, mode, language)
	stats := trainer.Run()
	if export != "" {
		ext := export[strings.LastIndex(export, ".")+1:]
		var data []byte
		var err error
		if ext == "json" {
			data, err = json.MarshalIndent(stats, "", "  ")
		} else if ext == "csv" {
			var line string
			for k, v := range stats {
				line += fmt.Sprintf("%v,", v)
			}
			line = line[:len(line)-1] + "\n"
			data = []byte(line)
		}
		if err == nil {
			os.WriteFile(export, data, 0644)
			fmt.Printf("Статистика сохранена в %s\n", export)
		}
	}
}

File: typing_trainer.cpp
// typing_trainer.cpp
#include <iostream>
#include <string>
#include <vector>
#include <random>
#include <chrono>
#include <thread>
#include <algorithm>
#include <ctime>
#include <iomanip>

using namespace std;

class Trainer {
private:
    int length;
    string charset, mode, language;
    string text;
    int errors;
    int totalChars;
    chrono::steady_clock::time_point start, end;

    string getCharset() {
        string letters = (language == "ru") ?
            "абвгдеёжзийклмнопрстуфхцчшщъыьэюяАБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ" :
            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
        string digits = "0123456789";
        string symbols = "!@#$%^&*()_+-=,.;:?/[]{}|";
        if (charset == "letters") return letters;
        if (charset == "alnum") return letters + digits;
        return letters + digits + symbols;
    }

    string generateText() {
        string cs = getCharset();
        random_device rd;
        mt19937 gen(rd());
        uniform_int_distribution<> dist(0, cs.size()-1);
        if (mode == "word") {
            int wordCount = max(1, length / 5);
            vector<string> words;
            for (int i = 0; i < wordCount; ++i) {
                int wl = 3 + (gen() % 6);
                string word;
                for (int j = 0; j < wl; ++j) word += cs[dist(gen)];
                words.push_back(word);
            }
            string result;
            for (size_t i = 0; i < words.size(); ++i) {
                if (i) result += " ";
                result += words[i];
            }
            return result;
        } else if (mode == "sentence") {
            int sentCount = max(1, length / 30);
            vector<string> sentences;
            for (int i = 0; i < sentCount; ++i) {
                int wc = 5 + (gen() % 8);
                vector<string> words;
                for (int j = 0; j < wc; ++j) {
                    int wl = 3 + (gen() % 6);
                    string word;
                    for (int k = 0; k < wl; ++k) word += cs[dist(gen)];
                    words.push_back(word);
                }
                string sent;
                for (size_t j = 0; j < words.size(); ++j) {
                    if (j) sent += " ";
                    sent += words[j];
                }
                if (!sent.empty()) sent[0] = toupper(sent[0]);
                sent += ".";
                sentences.push_back(sent);
            }
            string result;
            for (size_t i = 0; i < sentences.size(); ++i) {
                if (i) result += " ";
                result += sentences[i];
            }
            return result;
        } else {
            string result;
            for (int i = 0; i < length; ++i) result += cs[dist(gen)];
            return result;
        }
    }

public:
    Trainer(int len, const string& cs, const string& md, const string& lang)
        : length(len), charset(cs), mode(md), language(lang) {
        text = generateText();
        errors = 0;
        totalChars = 0;
    }

    void run() {
        cout << "\033[36mТренажер печати (символы)\033[0m" << endl;
        cout << "\033[33mВведите следующий текст как можно точнее и быстрее:\033[0m" << endl;
        cout << "\033[37m" << text << "\033[0m" << endl;
        cout << "\nНажмите Enter, чтобы начать...";
        cin.ignore();
        start = chrono::steady_clock::now();
        cout << "\033[32m> \033[0m";
        string input;
        getline(cin, input);
        end = chrono::steady_clock::now();
        totalChars = input.size();
        int minLen = min(text.size(), input.size());
        errors = 0;
        for (int i = 0; i < minLen; ++i) {
            if (text[i] != input[i]) ++errors;
        }
        if (input.size() > text.size()) errors += input.size() - text.size();
        double elapsed = chrono::duration<double>(end - start).count();
        double cpm = (totalChars / elapsed) * 60;
        double wpm = cpm / 5;
        double accuracy = totalChars > 0 ? (double)(totalChars - errors) / totalChars * 100 : 0;
        cout << "\n\033[36m--- Результаты ---\033[0m" << endl;
        cout << "\033[32mСкорость: " << fixed << setprecision(1) << cpm << " симв/мин (" << wpm << " слов/мин)\033[0m" << endl;
        cout << "\033[33mТочность: " << accuracy << "%\033[0m" << endl;
        cout << "\033[35mВремя: " << elapsed << " сек\033[0m" << endl;
        cout << "\033[31mОшибок: " << errors << "\033[0m" << endl;
    }
};

int main(int argc, char* argv[]) {
    int length = 50;
    string charset = "all", mode = "char", language = "en", exportFile;

    for (int i = 1; i < argc; ++i) {
        string arg = argv[i];
        if (arg == "--length" && i+1 < argc) length = stoi(argv[++i]);
        else if (arg == "--charset" && i+1 < argc) charset = argv[++i];
        else if (arg == "--mode" && i+1 < argc) mode = argv[++i];
        else if (arg == "--language" && i+1 < argc) language = argv[++i];
        else if (arg == "--export-stats" && i+1 < argc) exportFile = argv[++i];
    }

    Trainer trainer(length, charset, mode, language);
    trainer.run();
    // Экспорт в C++ не реализован для краткости
    return 0;
}
